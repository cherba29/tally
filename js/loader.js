goog.provide('budget.summary');
goog.provide('budget.summary.Loader');


goog.require('goog.dom');
goog.require('goog.events');
goog.require('goog.json');
goog.require('goog.net.EventType');
goog.require('goog.net.XhrIo');

goog.scope(function() {

budget.summary.Loader = function(budgetDataBuilder) {
  this.budgetDataBuilder_ = budgetDataBuilder;
}
var Loader = budget.summary.Loader;


/**
 * Loads json content as specified application path.
 * @param {string} contentPath path to budget data served by backend.
 * @param {function()}
 */
Loader.prototype.load = function(contentPath, fn) {
  var request = new goog.net.XhrIo();
  goog.events.listen(request, goog.net.EventType.COMPLETE, function() {
    if (request.isSuccess()) {
      console.log("Success");
      var data = request.getResponseJson();
      console.log("response", data);
      fn(data);
    } else {
      console.log("response error");
    }
  });
  request.send(
      '/budget?dir=' + contentPath,
      'GET',
      goog.json.serialize({}), 
      {'content-type':'application/json'});
}

budget.summary.hello = function() {
  var newDiv = goog.dom.createDom('h1', {'style': 'background-color:#EEE'},
    'Hello world2!');
  goog.dom.appendChild(document.body, newDiv);
};

});  // goog.scope

goog.exportSymbol('budget.summary.hello', budget.summary.hello);
goog.exportSymbol('budget.summary.Loader', budget.summary.Loader);
