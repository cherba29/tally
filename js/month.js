goog.provide('budget.Month');

goog.scope(function() {

/**
 * @param {number=} year Year for the month.
 * @param {number=} month Month [0-11] for this month instance. 
 * @final
 * @constructor
 */
budget.Month = function(year, month) {

};
var Account = budget.Account;

/**
 * @type {string}
 * @private
 */
Account.name_ = null;

/**
 * @type {string}
 * @private
 */
Account.description_ = null;

/**
 * @type {string}
 * @private
 */
Account.number_ = null;
Account.owners_ = [];

/**
 * @type {Month}
 * @private
 */
Account.openedOn_ = null;

/**
 * @type {Month}
 * @private
 */
Account.closedOn_ = null;

/**
 * @type {string}
 * @private
 */
Account.url_ = null;

/**
 * @type {string}
 * @private
 */
Account.userName_ = null;

/**
 * @type {string}
 * @private
 */
Account.password_ = null;

/**
 * @type {string}
 * @private
 */
Account.phone_ = null;

/**
 * @type {string}
 * @private
 */
Account.address_ = null;

});  // goog.scope

goog.exportSymbol('budget.Account', budget.Account);
