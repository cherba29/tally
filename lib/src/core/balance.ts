export enum Type {
  UNKNOWN = 'UNKNOWN',
  CONFIRMED = 'CONFIRMED',
  PROJECTED = 'PROJECTED',
  AUTO_PROJECTED = 'AUTO_PROJECTED'
}

const combineTypes = (t1: Type, t2: Type): Type => (t1 < t2 ? t2 : t1);

export class Balance {
  constructor(readonly amount: number, readonly date: Date, readonly type: Type) {}

  // Helper contructor.
  static confirmed(amount: number, date: string): Balance {
    return new Balance(amount, new Date(date), Type.CONFIRMED);
  }

  static projected(amount: number, date: string): Balance {
    return new Balance(amount, new Date(date), Type.PROJECTED);
  }

  static negated(balance: Balance): Balance {
    return new Balance(-balance.amount, balance.date, balance.type);
  }

  static add(balance1: Balance, balance2: Balance): Balance {
    const maxDate = balance1.date.getTime() < balance2.date.getTime() ? balance2.date : balance1.date;
    return new Balance(
      balance1.amount + balance2.amount,
      maxDate,
      combineTypes(balance1.type, balance2.type)
    );
  }

  static subtract(balance1: Balance, balance2: Balance): Balance {
    return Balance.add(balance1, new Balance(-balance2.amount, balance2.date, balance2.type));
  }

  compareTo(other: Balance): number {
    const dateDiff = this.date.getTime() - other.date.getTime();
    if (dateDiff !== 0) {
      return dateDiff;
    }
    const amountDiff = this.amount - other.amount;
    if (amountDiff !== 0) {
      return amountDiff;
    }
    return this.type.localeCompare(other.type);
  }

  toString(): string {
    return `Balance { amount: ${this.amount}, date: ${this.date
      .toISOString()
      .slice(0, 10)}, type: ${Type[this.type]} }`;
  }
}
