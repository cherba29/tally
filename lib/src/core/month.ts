const MONTH_NAMES = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
];
const MONTH_NAME_TO_INDEX = Object.fromEntries(MONTH_NAMES.map((x, i) => [x, i]));

/**
 * Represents a finincial month for accounting purposes.
 * Transcations from neighboring months can be attributed to it, for example
 * creadit card statement can close on 24th of the month, and for last week
 * of the month transactions will be attributed to next month.
 *
 * Month is represented as "MonthYear" string, for example "Mar2021".
 */
export class Month {
  constructor(readonly year: number, readonly month: number) {
    if (this.month < 0 || this.month > 11) {
      throw new Error(`Invalid value for month ${this.month}`);
    }
  }

  toString(): string {
    return `${MONTH_NAMES[this.month]}${this.year}`;
  }

  next(): Month {
    if (this.month >= 11) {
      return new Month(this.year + 1, 0);
    }
    return new Month(this.year, this.month + 1);
  }

  previous(): Month {
    if (this.month == 0) {
      return new Month(this.year - 1, 11);
    }
    return new Month(this.year, this.month - 1);
  }

  isLess(other: Month): boolean {
    if (this.year === other.year) {
      return this.month < other.month;
    }
    return this.year < other.year;
  }

  compareTo(other: Month): number {
    if (this.year === other.year) {
      return this.month - other.month;
    }
    return this.year - other.year;
  }

  /**
   * Number of months between two month dates.
   * Negative is provided date is larger
   */
  distance(other: Month): number {
    return (this.year - other.year) * 12 + (this.month - other.month);
  }

  /** Convert string representation to internal representation. */
  static fromString(name: string): Month {
    if (name.length < 4) {
      throw new Error(`Cant get month from small string "${name}"`);
    }
    const month = MONTH_NAME_TO_INDEX[name.substring(0, 3)];
    if (month === undefined) {
      throw new Error(`Cant find month for "${name}"`);
    }
    const year = Number(name.substring(3));
    if (isNaN(year)) {
      throw Error(`Cant get year from "${name}"`);
    }
    return new Month(year, month);
  }

  static fromDate(date: Date): Month {
    return new Month(date.getUTCFullYear(), date.getUTCMonth());
  }

  /** Creates generator spanning start and end (but not including) months. */
  static *generate(start: Month, end: Month): IterableIterator<Month> {
    let current = start;
    while (current.isLess(end)) {
      yield current;
      current = current.next();
    }
    return current;
  }
}
