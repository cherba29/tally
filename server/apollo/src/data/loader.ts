import * as fs from 'fs';
import * as path from 'path';

import { Budget, BudgetBuilder } from '@tally/lib';
import { loadYamlFile } from './loader_yaml';
import { loadTallyConfig } from './config';
// import {SummaryStatement} from '../core/summary_statement';

function* readdirSync(file_path: string): Generator<string> {
  if (fs.statSync(file_path).isDirectory()) {
    for (const f of fs.readdirSync(file_path)) {
      const child_file = path.join(file_path, f);
      for (const cf of readdirSync(child_file)) {
        if (cf.match(/\.yaml$/)) {
          yield cf;
        }
      }
    }
  } else {
    yield file_path;
  }
  return;
}

export function listFiles(): string[] {
  if (!process.env.TALLY_FILES) {
    throw Error('Process environment variable "TALLY_FILES" has not been specified.');
  }
  const filePaths: string[] = [];
  for (const file_path of readdirSync(process.env.TALLY_FILES)) {
    const relative_file_path = file_path.slice(process.env.TALLY_FILES.length + 1);
    filePaths.push(relative_file_path);
  }
  return filePaths;
}

export function loadBudget(): Budget {
  if (!process.env.TALLY_FILES) {
    throw Error('Process environment variable "TALLY_FILES" has not been specified.');
  }
  const budgetBuilder = new BudgetBuilder();

  const config = loadTallyConfig();
  // Always show one more month, in case user chooses in the config
  // just single month. It is not meaningful to show only one month
  // as values from next used for current month.
  budgetBuilder.setPeriod(config.budget_period.start, config.budget_period.end.next());

  const filePaths: string[] = [];
  for (const file_path of readdirSync(process.env.TALLY_FILES)) {
    const relative_file_path = file_path.slice(process.env.TALLY_FILES.length + 1);
    filePaths.push(relative_file_path);
    console.log('Loading ' + relative_file_path);
    const content = fs.readFileSync(file_path, 'utf8');
    loadYamlFile(budgetBuilder, content, relative_file_path);
  }
  console.log(`Done loading ${filePaths.length} file(s)`);
  return budgetBuilder.build();
}
