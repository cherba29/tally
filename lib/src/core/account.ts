import { Month } from './month';

export enum Type {
  UNSPECIFIED = '_unspecified_',
  BILL = 'bill',
  CHECKING = 'checking',
  CREDIT = 'credit',
  CREDIT_CARD = 'credit-card',
  DEFERRED_INCOME = 'deferred income',
  EXTERNAL = 'external',
  INCOME = 'income',
  INVESTMENT = 'investment',
  RETIREMENT = 'retirement',
  SUMMARY = '_summary_',
  TAX = 'tax_',
}

// Typescript cannot reverse map string enums, so prebuild a lookup map.
// https://www.typescriptlang.org/docs/handbook/enums.html
const reverseAccountType = new Map<string, string>(Object.entries(Type).map(([k, v]) => [v, k]));

export interface InitData {
  name: string;
  description?: string;
  parents?: Account[];
  children?: Account[];
  autoBalance?: boolean;
  type: Type;
  number?: string;
  openedOn?: Month;
  closedOn?: Month;
  owners: string[];
  url?: string;
  address?: string;
  phone?: string;
  userName?: string;
  password?: string;
}

export class Account {
  readonly name: string;
  readonly description?: string;
  readonly parents: Account[];
  readonly children: Account[];
  // Account type, for example 'CREDIT_CARD'.
  readonly type: Type;
  // Account balances are automatically considered to be committed.
  readonly autoBalance: boolean;
  // Real account number associated with this account.
  readonly number?: string;
  // Month when account was opened and possibly closed.
  readonly openedOn?: Month;
  readonly closedOn?: Month;
  // List of account owner ids.
  readonly owners: string[];
  // Url to the account.
  readonly url?: string;
  // Physical address for the account.
  readonly address?: string;
  // Phone number for customer support.
  readonly phone?: string;
  // Username/password to use to login to the account.
  readonly userName?: string;
  readonly password?: string;

  constructor(data: InitData) {
    this.name = data.name;
    this.description = data.description;
    this.parents = data.parents ?? [];
    this.children = data.children ?? [];
    this.autoBalance = data.autoBalance ?? false;
    this.type = data.type;
    this.number = data.number;
    this.openedOn = data.openedOn;
    this.closedOn = data.closedOn;
    this.owners = data.owners;
    this.url = data.url;
    this.address = data.address;
    this.phone = data.phone;
    this.userName = data.userName;
    this.password = data.password;
  }

  toString() {
    return `Account ${this.name} ${this.type}${this.closedOn ? ` Closed ${this.closedOn}` : ''}`;
  }

  isClosed(month: Month): boolean {
    return !!(
      (
        this.closedOn?.isLess(month) || // After closed.
        (this.openedOn && month.isLess(this.openedOn))
      ) // Before or on open.
    );
  }

  // Is this an external/bookeeping account. External accounts are not
  // considered to be part of any owner.
  get isExternal(): boolean {
    return (
      this.type === Type.EXTERNAL || this.type === Type.TAX || this.type === Type.DEFERRED_INCOME
    );
  }

  // Is this synthetic summary account.
  get isSummary(): boolean {
    return this.type === Type.SUMMARY;
  }

  hasCommonOwner(other: Account): boolean {
    return this.owners.some((owner) => other.owners.includes(owner));
  }

  get typeIdName(): string {
    return reverseAccountType.get(this.type.toString()) || '';
  }
}
