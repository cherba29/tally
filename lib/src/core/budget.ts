import { Account } from './account';
import { Balance } from './balance';
import { Month } from './month';
import { Transfer } from './transfer';

export class Budget {
  constructor(
    // Period over which the budget is defined.
    readonly months: Month[],
    // Account name to account map.
    readonly accounts: Map<string, Account>,
    // Account name -> month -> balance map.
    readonly balances: Map<string, Map<string, Balance>>,
    // Account name -> month -> transfers map.
    readonly transfers: Map<string, Map<string, Set<Transfer>>>
  ) {}

  findActiveAccounts(): Account[] {
    const accounts: Account[] = [];
    for (const account of this.accounts.values()) {
      if (this.months.some((m) => !account.isClosed(m))) {
        accounts.push(account);
      }
    }
    return accounts;
  }
}

export interface TransferData {
  toAccount: string;
  toMonth: Month;
  fromAccount: string;
  fromMonth: Month;
  balance: Balance;
  description?: string;
}

function getMonthTransfers<T>(
  transfers: Map<string, Map<string, Set<T>>>,
  accountName: string,
  month: string
) {
  let accountTransfers = transfers.get(accountName);
  if (!accountTransfers) {
    accountTransfers = new Map<string, Set<T>>();
    transfers.set(accountName, accountTransfers);
  }
  let monthTransfers = accountTransfers.get(month);
  if (!monthTransfers) {
    monthTransfers = new Set<T>();
    accountTransfers.set(month, monthTransfers);
  }
  return monthTransfers;
}

export class BudgetBuilder {
  // Period over which the budget is defined.
  months: Month[] = [];
  // Account name to account map.
  readonly accounts = new Map<string, Account>();
  // Account name -> month -> balance map.
  readonly balances = new Map<string, Map<string, Balance>>();
  readonly transfers: TransferData[] = [];

  setPeriod(start: Month, end: Month): void {
    this.months = Array.from(Month.generate(start, end));
  }

  setAccount(account: Account): void {
    this.accounts.set(account.name, account);
  }

  setBalance(accountName: string, month: string, balance: Balance): void {
    let balances = this.balances.get(accountName);
    if (!balances) {
      balances = new Map<string, Balance>();
      this.balances.set(accountName, balances);
    }
    const existingBalance = balances.get(month);
    if (existingBalance) {
      throw new Error(
        `Balance for '${accountName}' '${month}' is already set to ${balance.toString()}`
      );
    }
    balances.set(month, balance);
  }

  addTransfer(transferData: TransferData): void {
    this.transfers.push(transferData);
  }

  build(): Budget {
    const transfers = new Map<string, Map<string, Set<Transfer>>>();
    for (const transferData of this.transfers) {
      const toAccount = this.accounts.get(transferData.toAccount);
      if (!toAccount) {
        throw new Error(`Unknown account ${transferData.toAccount}`);
      }
      const fromAccount = this.accounts.get(transferData.fromAccount);
      if (!fromAccount) {
        throw new Error(`Unknown account ${transferData.fromAccount}`);
      }
      const transfer: Transfer = {
        toAccount,
        fromAccount,
        toMonth: transferData.toMonth,
        fromMonth: transferData.fromMonth,
        balance: transferData.balance,
        description: transferData.description
      };
      const toMonthTransfers = getMonthTransfers(
        transfers,
        toAccount.name,
        transferData.toMonth.toString()
      );
      toMonthTransfers.add(transfer);
      const fromMonthTransfers = getMonthTransfers(
        transfers,
        fromAccount.name,
        transferData.fromMonth.toString()
      );
      fromMonthTransfers.add(transfer);
    }
    return new Budget(this.months, this.accounts, this.balances, transfers);
  }
}
