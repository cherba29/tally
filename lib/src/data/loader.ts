import * as fs from 'fs';
import * as path from 'path';
import * as chokidar from 'chokidar';

import { Budget, BudgetBuilder } from '../core/budget';
import { loadYamlFile, parseYamlContent, YamlData } from './loader_yaml';
import { TransactionStatement, buildTransactionStatementTable } from '../statement/transaction';
import { SummaryStatement, buildSummaryStatementTable } from '../statement/summary';
import { Map3 } from '../utils';

function* readdirSync(file_path: string): Generator<string> {
  if (fs.statSync(file_path).isDirectory()) {
    for (const f of fs.readdirSync(file_path)) {
      const child_file = path.join(file_path, f);
      for (const cf of readdirSync(child_file)) {
        if (cf.match(/\.yaml$/)) {
          yield cf;
        }
      }
    }
  } else {
    yield file_path;
  }
  return;
}

async function realPath(path: string): Promise<string> {
  const stats = await fs.promises.lstat(path);
  return stats.isSymbolicLink() ? await fs.promises.readlink(path) : path;
}

export function listFiles(): string[] {
  if (!process.env.TALLY_FILES) {
    throw Error('Process environment variable "TALLY_FILES" has not been specified.');
  }
  const filePaths: string[] = [];
  for (const file_path of readdirSync(process.env.TALLY_FILES)) {
    const relative_file_path = file_path.slice(process.env.TALLY_FILES.length + 1);
    filePaths.push(relative_file_path);
  }
  return filePaths;
}

let watchedPath: string | undefined = undefined;
let watcher: chokidar.FSWatcher | undefined = undefined;
let processedBudget: ProcessedBudget | undefined = undefined;

export function unwatchBudgetFiles() {
  watchedPath = undefined;
  watcher?.close();
  watcher = undefined;
  processedBudget = undefined;
}

export interface DataPayload {
  budget: Budget;
  statements: Map<string, Map<string, TransactionStatement>>;
  // owner -> accout name -> month -> summary.
  summaries: Map3<SummaryStatement>;
}

class ProcessedBudget {
  private readonly parsedAccountData = new Map<string, YamlData>();
  budget?: Budget = undefined;
  readonly accoutToMonthToTransactionStatement = new Map<string, Map<string, TransactionStatement>>();
  summaryNameMonthMap = new Map3<SummaryStatement>();

  reProcess() {
    const startTimeMs: number = Date.now();
    const budgetBuilder = new BudgetBuilder();
    for (const [filePath, accountData] of this.parsedAccountData) {
      try {
        loadYamlFile(budgetBuilder, accountData, filePath);
        if (!budgetBuilder.accounts.has(accountData?.name ?? '')) {
          console.warn(`warning: ${filePath} is not an account file.`);
        }
      } catch (e) {
        console.error(`error: Failed to add ${filePath}, ${e}`);
        return;
      }
    }
    this.budget = budgetBuilder.build();
    console.log(`Done building budget for ${this.budget.accounts.size} accounts ` 
        + ` in ${Date.now() - startTimeMs}ms`);

    this.accoutToMonthToTransactionStatement.clear();
    this.summaryNameMonthMap.clear();

    const startBuildTransactionStatementsTimeMs: number = Date.now();
    const transactionStatementTable = buildTransactionStatementTable(this.budget);
    for (const stmt of transactionStatementTable) {
      let monthToStatement = this.accoutToMonthToTransactionStatement.get(stmt.account.name);
      if (!monthToStatement) {
        monthToStatement = new Map<string, TransactionStatement>();
        this.accoutToMonthToTransactionStatement.set(stmt.account.name, monthToStatement);
      }
      monthToStatement.set(stmt.month.toString(), stmt);
    }
    console.log(`Done building ${transactionStatementTable.length} transaction statements ` 
                + ` in ${Date.now() - startBuildTransactionStatementsTimeMs}ms`);

    const startBuildSummaryStatementsTimeMs: number = Date.now();
    this.summaryNameMonthMap = buildSummaryStatementTable(transactionStatementTable);
    let numSummaryStatements = this.summaryNameMonthMap.size;
    console.log(`Done building ${numSummaryStatements} summary statements ` 
                + ` in ${Date.now() - startBuildSummaryStatementsTimeMs}ms`);

    console.log(`Done reprocessing ${this.parsedAccountData.size} file(s) ` 
        + `${transactionStatementTable.length} tran statements and ` 
        + `${numSummaryStatements} summaries in ${Date.now() - startTimeMs}ms`);
  }

  addFolder(rootPath: string) {
    const startTimeMs: number = Date.now();
    for (const filePath of readdirSync(rootPath)) {
      const relativeFilePath = filePath.slice(rootPath.length + 1);
      console.log('Loading ' + relativeFilePath);
      this.addFile(rootPath, relativeFilePath);
    }
    console.log(`Done loading ${this.parsedAccountData.size} file(s) in ${Date.now() - startTimeMs}ms`);
  }

  addFile(rootPath: string, relativeFilePath: string) {
    const content = fs.readFileSync(`${rootPath}/${relativeFilePath}`, 'utf8');
    const accountData = parseYamlContent(content, relativeFilePath);
    if (accountData === undefined) {
      throw Error(
        `Failed to parse ${relativeFilePath} content of size ${content.length} fileStat: ${relativeFilePath}`
      );
    }
    this.parsedAccountData.set(relativeFilePath, accountData);
  }
}

export async function loadBudget(): Promise<DataPayload> {
  if (processedBudget !== undefined) {
    return { 
      budget: processedBudget.budget!, 
      statements: processedBudget.accoutToMonthToTransactionStatement, 
      summaries: processedBudget.summaryNameMonthMap,  
    };
  
  }
  if (!process.env.TALLY_FILES) {
    throw Error('Process environment variable "TALLY_FILES" has not been specified.');
  }
  
  processedBudget = new ProcessedBudget();
  
  const pathToData = await realPath(process.env.TALLY_FILES);
  processedBudget.addFolder(pathToData);
  processedBudget.reProcess();

  if (watcher === undefined || watchedPath !== pathToData) {
    if (watcher) {
      unwatchBudgetFiles();
    }
    watchedPath = pathToData;
    watcher = chokidar
      .watch(pathToData, {
        persistent: true,
        // Multiple change events can be emitted as a larger file is being written.
        awaitWriteFinish: {
          stabilityThreshold: 300,
        },
      })
      .on('all', async (eventType: string, fileStat: fs.Stats | undefined) => {
        if (!fileStat || ['add', 'addDir'].includes(eventType)) {
          return;
        }
        console.log(`${eventType} for ${fileStat}`);
        const startTimeMs: number = Date.now();
        const relativeFilePath = `${fileStat}`.slice(pathToData.length + 1);
        if (processedBudget) {
          processedBudget.addFile(pathToData, relativeFilePath);
          processedBudget.reProcess();
        } else {
          console.error(`Unexpected error: processedBudget is unset`);
        }
        console.log(`info: Rebuilt budget in ${Date.now() - startTimeMs}ms`);
      });
    console.log(`info: Watching ${process.env.TALLY_FILES}`);
  }
  return { 
    budget: processedBudget.budget!, 
    statements: processedBudget.accoutToMonthToTransactionStatement, 
    summaries:  processedBudget.summaryNameMonthMap,  
  };
}
