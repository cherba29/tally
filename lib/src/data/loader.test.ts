import { afterEach, describe, expect, jest, test } from '@jest/globals';
import { listFiles, loadBudget } from './loader';
import mockfs from 'mock-fs';
import { Month } from '../core/month';

describe('listFiles', () => {
  afterEach(() => {
    mockfs.restore();
    delete process.env.TALLY_FILES;
  });

  test('fails if environment not set', () => {
    expect(listFiles).toThrow(
      new Error('Process environment variable "TALLY_FILES" has not been specified.')
    );
  });

  test('returns empty on empty directory', () => {
    const TALLY_PATH = 'tally/files/path';
    process.env.TALLY_FILES = TALLY_PATH;
    mockfs({
      [TALLY_PATH]: {
        /* empty directory */
      },
    });
    expect(listFiles()).toEqual([]);
  });

  test('returns recursively yaml file entries', () => {
    const TALLY_PATH = 'tally/files/path';
    process.env.TALLY_FILES = TALLY_PATH;
    mockfs({
      [TALLY_PATH]: {
        '_config.yaml': '',
        subdir1: {
          'file1.json': '',
          'file2.yaml': '',
        },
        subdir2: {
          'file1.json': '',
          'file2.yaml': '',
        },
      },
    });
    expect(listFiles()).toEqual(['_config.yaml', 'subdir1/file2.yaml', 'subdir2/file2.yaml']);
  });
});

describe('loadBudget', () => {
  afterEach(() => {
    mockfs.restore();
    delete process.env.TALLY_FILES;
    jest.clearAllMocks();
  });

  test('fails if environment path is not set', () => {
    expect(loadBudget).toThrow(
      new Error('Process environment variable "TALLY_FILES" has not been specified.')
    );
  });

  test('fails without config', () => {
    const TALLY_PATH = 'tally/files/path';
    process.env.TALLY_FILES = TALLY_PATH;
    mockfs({
      [TALLY_PATH]: {
        /* empty directory */
      },
    });
    expect(loadBudget).toThrow(
      new Error("ENOENT: no such file or directory, open 'tally/files/path/_config.yaml'")
    );
  });

  test('empty', () => {
    jest.spyOn(console, 'log').mockImplementation(() => {});

    const TALLY_PATH = 'tally/files/path';
    process.env.TALLY_FILES = TALLY_PATH;
    mockfs({
      [TALLY_PATH]: {
        '_config.yaml': 'budget_period: {start: Nov2019, end: Feb2020}',
      },
    });
    expect(loadBudget()).toEqual({
      accounts: new Map(),
      balances: new Map(),
      months: [new Month(2019, 10), new Month(2019, 11), new Month(2020, 0), new Month(2020, 1)],
      transfers: new Map(),
    });

    expect(console.log).toHaveBeenCalledTimes(2);
    expect(console.log).toHaveBeenCalledWith('Loading _config.yaml');
    expect(console.log).toHaveBeenCalledWith('Done loading 1 file(s)');
  });
});
