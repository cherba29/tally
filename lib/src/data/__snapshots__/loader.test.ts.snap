// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadBudget just account 1`] = `
{
  "budget": Budget {
    "accounts": Map {
      "test-account" => Account {
        "address": undefined,
        "closedOn": undefined,
        "description": undefined,
        "name": "test-account",
        "number": undefined,
        "openedOn": Month {
          "month": 2,
          "year": 2019,
        },
        "owners": [
          "someone",
        ],
        "password": undefined,
        "path": [],
        "phone": undefined,
        "type": "external",
        "url": undefined,
        "userName": undefined,
      },
    },
    "balances": Map {
      "test-account" => Map {
        "Mar2019" => Balance {
          "amount": 10000,
          "date": 2019-03-01T00:00:00.000Z,
          "type": "CONFIRMED",
        },
      },
    },
    "months": [
      Month {
        "month": 2,
        "year": 2019,
      },
    ],
    "transfers": Map {},
  },
  "statements": Map {
    "test-account" => Map {
      "Mar2019" => TransactionStatement {
        "account": Account {
          "address": undefined,
          "closedOn": undefined,
          "description": undefined,
          "name": "test-account",
          "number": undefined,
          "openedOn": Month {
            "month": 2,
            "year": 2019,
          },
          "owners": [
            "someone",
          ],
          "password": undefined,
          "path": [],
          "phone": undefined,
          "type": "external",
          "url": undefined,
          "userName": undefined,
        },
        "coversPrevious": false,
        "coversProjectedPrevious": false,
        "endBalance": undefined,
        "hasProjectedTransfer": false,
        "inFlows": 0,
        "income": 0,
        "isCovered": true,
        "isProjectedCovered": true,
        "month": Month {
          "month": 2,
          "year": 2019,
        },
        "outFlows": 0,
        "startBalance": Balance {
          "amount": 10000,
          "date": 2019-03-01T00:00:00.000Z,
          "type": "CONFIRMED",
        },
        "totalPayments": 0,
        "totalTransfers": 0,
        "transactions": [],
      },
    },
  },
  "summaries": Map3 {
    "store": Map {
      "someone" => Map {
        "someone EXTERNAL" => Map {
          "Mar2019" => SummaryStatement {
            "account": Account {
              "address": undefined,
              "closedOn": undefined,
              "description": undefined,
              "name": "someone EXTERNAL",
              "number": undefined,
              "openedOn": undefined,
              "owners": [
                "someone",
              ],
              "password": undefined,
              "path": [],
              "phone": undefined,
              "type": "_summary_",
              "url": undefined,
              "userName": undefined,
            },
            "endBalance": undefined,
            "inFlows": 0,
            "income": 0,
            "month": Month {
              "month": 2,
              "year": 2019,
            },
            "outFlows": 0,
            "startBalance": Balance {
              "amount": 10000,
              "date": 2019-03-01T00:00:00.000Z,
              "type": "CONFIRMED",
            },
            "startMonth": Month {
              "month": 2,
              "year": 2019,
            },
            "statements": [
              TransactionStatement {
                "account": Account {
                  "address": undefined,
                  "closedOn": undefined,
                  "description": undefined,
                  "name": "test-account",
                  "number": undefined,
                  "openedOn": Month {
                    "month": 2,
                    "year": 2019,
                  },
                  "owners": [
                    "someone",
                  ],
                  "password": undefined,
                  "path": [],
                  "phone": undefined,
                  "type": "external",
                  "url": undefined,
                  "userName": undefined,
                },
                "coversPrevious": false,
                "coversProjectedPrevious": false,
                "endBalance": undefined,
                "hasProjectedTransfer": false,
                "inFlows": 0,
                "income": 0,
                "isCovered": true,
                "isProjectedCovered": true,
                "month": Month {
                  "month": 2,
                  "year": 2019,
                },
                "outFlows": 0,
                "startBalance": Balance {
                  "amount": 10000,
                  "date": 2019-03-01T00:00:00.000Z,
                  "type": "CONFIRMED",
                },
                "totalPayments": 0,
                "totalTransfers": 0,
                "transactions": [],
              },
            ],
            "totalPayments": 0,
            "totalTransfers": 0,
          },
        },
      },
    },
  },
}
`;
